### ISSUE TITLE
CMake build from source is broken

### ISSUE THREAD

**Issue by mofeing** [QE-Lab/dqcsim#646]
For the past week I have been trying to build my own DQCsim backend in C++. I have tried to install first DQCsim into a virtualenv using `pip` and also to install it via FetchContent with `DQCSIM_FROM_SOURCE` equal to `yes`, but there is always an error happening.

I'm trying to build the example in `examples/cmake-fetchcontent` with `DQCSIM_FROM_SOURCE` equal to `yes` but it gives an error:

```
Scanning dependencies of target dqcsim_build
[ 10%] Creating directories for 'dqcsim_build'
[ 20%] No download step for 'dqcsim_build'
[ 30%] No update step for 'dqcsim_build'
[ 40%] No patch step for 'dqcsim_build'
[ 50%] Performing configure step for 'dqcsim_build'
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /Users/mofeing/.pyenv/shims/python (found version "3.8.2") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Failed to find LLVM FileCheck
-- Found Git: /usr/local/bin/git (found version "2.29.2") 
-- git Version: v1.5.2-348aa2c9
-- Version: 1.5.2
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Success
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Success
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Success
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED - Success
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Success
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES -- failed to compile
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK -- success
-- Using the single-header code from /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/dqcsim_build-prefix/src/dqcsim_build-build/_deps/json-src/single_include/
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/dqcsim_build-prefix/src/dqcsim_build-build
[ 60%] Performing build step for 'dqcsim_build'
Scanning dependencies of target dqcsim_crate
[ 12%] Creating directories for 'dqcsim_crate'
[ 25%] No download step for 'dqcsim_crate'
[ 37%] No update step for 'dqcsim_crate'
[ 50%] No patch step for 'dqcsim_crate'
[ 62%] No configure step for 'dqcsim_crate'
[ 75%] Performing build step for 'dqcsim_crate'
   Compiling proc-macro2 v1.0.18
   Compiling libc v0.2.71
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.33
   Compiling cfg-if v0.1.10
   Compiling serde_derive v1.0.114
   Compiling serde v1.0.114
   Compiling getrandom v0.1.14
   Compiling autocfg v1.0.0
   Compiling ppv-lite86 v0.2.8
   Compiling lazy_static v1.4.0
   Compiling ryu v1.0.5
   Compiling bitflags v1.2.1
   Compiling memchr v2.3.3
   Compiling log v0.4.10
   Compiling unicode-width v0.1.8
   Compiling remove_dir_all v0.5.3
   Compiling unicode-segmentation v1.6.0
   Compiling serde_json v1.0.56
   Compiling ansi_term v0.11.0
   Compiling vec_map v0.8.2
   Compiling itoa v0.4.6
   Compiling strsim v0.8.0
   Compiling cbindgen v0.14.3
   Compiling maybe-uninit v2.0.0
   Compiling byteorder v1.3.4
   Compiling regex-syntax v0.6.18
   Compiling failure_derive v0.1.8
   Compiling adler32 v1.1.0
   Compiling gimli v0.21.0
   Compiling object v0.20.0
   Compiling linked-hash-map v0.5.3
   Compiling rustc-demangle v0.1.16
   Compiling fnv v1.0.7
   Compiling dtoa v0.4.6
   Compiling half v1.6.0
   Compiling is_executable v0.1.2
   Compiling whoami v0.9.0
   Compiling integer-sqrt v0.1.3
   Compiling pathdiff v0.2.0
   Compiling strum v0.18.0
   Compiling humantime v2.0.1
   Compiling ref_thread_local v0.0.0
   Compiling thread_local v1.0.1
   Compiling num-traits v0.2.12
   Compiling crossbeam-utils v0.7.2
   Compiling textwrap v0.11.0
   Compiling heck v0.3.1
   Compiling miniz_oxide v0.3.7
   Compiling yaml-rust v0.4.4
   Compiling addr2line v0.12.2
   Compiling quote v1.0.2
   Compiling atty v0.2.14
   Compiling dirs-sys v0.3.5
   Compiling backtrace v0.3.49
   Compiling aho-corasick v0.7.13
   Compiling rand_core v0.5.1
   Compiling clap v2.33.1
   Compiling dirs v2.0.2
   Compiling crossbeam-channel v0.4.2
   Compiling rand_chacha v0.2.2
   Compiling num-complex v0.3.0
   Compiling float-cmp v0.8.0
   Compiling regex v1.3.9
   Compiling term v0.6.1
   Compiling rand v0.7.3
   Compiling tempfile v3.1.0
   Compiling uuid v0.8.1
   Compiling synstructure v0.12.4
   Compiling strum_macros v0.18.0
   Compiling failure v0.1.8
   Compiling toml v0.5.6
   Compiling bincode v1.3.1
   Compiling serde_yaml v0.8.13
   Compiling serde_cbor v0.11.1
   Compiling serde-transcode v1.1.0
   Compiling ipc-channel v0.14.1
   Compiling dqcsim v0.2.4 (/Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-src/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 34s
[ 87%] No install step for 'dqcsim_crate'
[100%] Completed 'dqcsim_crate'
[100%] Built target dqcsim_crate
[ 70%] Performing install step for 'dqcsim_build'
[ 12%] Performing build step for 'dqcsim_crate'
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
[ 25%] No install step for 'dqcsim_crate'
[ 37%] Completed 'dqcsim_crate'
[100%] Built target dqcsim_crate
Install the project...
-- Install configuration: "debug"
-- Up-to-date: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim_bits
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim_bits/wrap.hpp
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim_bits/arb_builder.hpp
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim_bits/log.hpp
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/cdqcsim
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/include/dqcsim.h
-- Installing: /Users/mofeing/Develop/dqcsim/examples/cmake-fetchcontent/build/_deps/dqcsim-build/_dqcsim_install_prefix_/lib/libdqcsim.dylib
[ 80%] Completed 'dqcsim_build'
[ 80%] Built target dqcsim_build
Scanning dependencies of target main
make[2]: *** No rule to make target `_deps/dqcsim-build/_dqcsim_install_prefix_/lib64/libdqcsim.dylib', needed by `main'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 90%] Building CXX object CMakeFiles/main.dir/main.cpp.o
make[1]: *** [CMakeFiles/main.dir/all] Error 2
make: *** [all] Error 2
```

I have followed the error to [CMakeLists.txt@38](https://github.com/QE-Lab/dqcsim/blob/821e4bace6061b05db7b1e62104a39e1dd87c304/CMakeLists.txt#L38), due to the fact that `libdqcsim.dylib` resides in `_deps/dqcsim-build/_dqcsim_install_prefix_/lib` and **not** `_deps/dqcsim-build/_dqcsim_install_prefix_/lib64` in my computer (macos 10.15).

I would say that the error is to try to hardcode the path to the shared library instead of trying to use `find_library`. Like this, there are other lines of code that are bad CMake practice like for example the use of `if(PROJECT_NAME)` (when you should not care about being imported or not) or `if(DQCSIM_FROM_SOURCE)` when you could use `option` or `set(... CACHE ...)`.

We are interested in trying DQCsim so I can lend you a hand. I want to refactor the [CMakeLists.txt](https://github.com/QE-Lab/dqcsim/blob/master/CMakeLists.txt) to use the so-called Modern CMake practices (and so, to be less error-prone) but I wanted first to ask for your permision and if there is anything I should take into account.

**Comment by mbrobbel** [comment]
> We are interested in trying DQCsim so I can lend you a hand. I want to refactor the CMakeLists.txt to use the so-called Modern CMake practices (and so, to be less error-prone) but I wanted first to ask for your permision and if there is anything I should take into account.

Yes please! Feel free to suggest improvements. I've added an example and fixed your issue in #647.

**Comment by mofeing** [comment]
Great! I'm going to code a proof-of-concept backend and then I will get my hands into this.
